A74-xds-config-tears.md

```diff
@@ -0,0 +1,346 @@
+A74: xDS Config Tears
+----
+* Author(s): @markdroth
+* Approver: @ejona86
+* Status: {Draft, In Review, Ready for Implementation, Implemented}
+* Implemented in: <language, ...>
+* Last updated: 2023-12-28
+* Discussion at: https://groups.google.com/g/grpc-io/c/ifcC3DbopWM
+
+## Abstract
+
+xDS configuration is split across four main resource types: listener,
+route configuration, cluster, and endpoint.  Today, these different
+resource types are watched in different parts of the gRPC channel:
+the listener and route configuration are watched in the xds resolver,
+whereas the cluster and endpoint resources are watched in the LB policy
+tree (in the cds and xds_cluster_resolver LB policies, respectively).
+This split prevents us from applying config changes to the channel in
+an atomic way, meaning that we can see config tears, which can lead
+to unnecessary latency and even spurious RPC failures.  This document
+proposes a design to fix that problem.
+
+## Background
+
+The way that we currently split the xDS resource watches between the
+resolver and the LB policy tree means that we cannot apply config updates
+to the gRPC channel as an atomic whole.  However, there are cases where a
+control plane needs to send a config change that spans multiple resources.
+
+At minimum, this can cause us to impose unnecessary latency for RPCs.
+For example, the xds resolver may apply a new route configuration that
+references a new cluster, even though the client has not yet obtained
+the resource for the new cluster.  As soon as the xds resolver applies
+that change, the ConfigSelector will start using the new route
+configuration, which means that it will start sending RPCs to the new
+cluster.  However, those RPCs may be delayed until the cds LB policy
+obtains the cluster resource.
+
+In some edge cases, this can even cause us to incorrectly fail RPCs.
+For example, consider a case using xDS fallback (see [A71]), where the
+primary server and the fallback server use different resource names
+for the cluster resources.  If the client is using the data from the
+fallback server and then the primary server comes back up, the client
+may see the does-not-exist notification for the cluster name from the
+fallback server before it sees the new route configuration that points
+to the correct cluster name, thus causing us to temporarily fail RPCs
+routed to that cluster.
+
+This document proposes a design to fix this by moving all of the
+resource watches into the xds resolver.
+
+### Related Proposals: 
+* [A27: xDS Global Load Balancing][A27]
+* [A28: xDS Traffic Splitting and Routing][A28]
+* [A29: xDS mTLS Security][A29]
+* [A31: xDS Timeout Support and Config Selector Design][A31]
+* [A37: xDS Aggregate and Logical DNS Clusters][A37]
+* [A55: xDS-Based Stateful Session Affinity for Proxyless gRPC][A55]
+* [A56: Priority LB Policy][A56]
+* [A71: xDS Fallback (pending)][A71]
```

Comment:
Can be removed from pending status and not link to pull.

Reply:
Done.